tests/CredentialsLoaderTest.php

```diff
@@ -29,6 +29,123 @@ public function testUpdateMetadataSkipsWhenAuthenticationisSet()
         $this->assertArrayHasKey('authentication', $metadata);
         $this->assertEquals('foo', $metadata['authentication']);
     }
+
+    /** @runInSeparateProcess */
+    public function testGetDefaultClientCertSource()
+    {
+        putenv('HOME=' . __DIR__ . '/fixtures4/valid');
+
+        $callback = CredentialsLoader::getDefaultClientCertSource();
+        $this->assertNotNull($callback);
+
+        $output = $callback();
+        $this->assertEquals('foo', $output);
+    }
+
+    /** @runInSeparateProcess */
+    public function testNonExistantDefaultClientCertSource()
+    {
+        putenv('HOME=');
+
+        $callback = CredentialsLoader::getDefaultClientCertSource();
+        $this->assertNull($callback);
+    }
+
+    /**
+     * @runInSeparateProcess
+     * @expectedException UnexpectedValueException
+     * @expectedExceptionMessage Invalid client cert source JSON
+     */
+    public function testDefaultClientCertSourceInvalidJsonThrowsException()
+    {
+        putenv('HOME=' . __DIR__ . '/fixtures4/invalidjson');
+
+        CredentialsLoader::getDefaultClientCertSource();
+    }
+
+    /**
+     * @runInSeparateProcess
+     * @expectedException UnexpectedValueException
+     * @expectedExceptionMessage cert source requires "cert_provider_command"
+     */
+    public function testDefaultClientCertSourceInvalidKeyThrowsException()
+    {
+        putenv('HOME=' . __DIR__ . '/fixtures4/invalidkey');
+
+        CredentialsLoader::getDefaultClientCertSource();
+    }
+
+    /**
+     * @runInSeparateProcess
+     * @expectedException UnexpectedValueException
+     * @expectedExceptionMessage cert source expects "cert_provider_command" to be an array
+     */
+    public function testDefaultClientCertSourceInvalidValueThrowsException()
+    {
+        putenv('HOME=' . __DIR__ . '/fixtures4/invalidvalue');
+
+        CredentialsLoader::getDefaultClientCertSource();
+    }
+
+    /**
+     * @runInSeparateProcess
+     */
+    public function testActualDefaultClientCertSource()
+    {
+        $clientCertSource = CredentialsLoader::getDefaultClientCertSource();
+        if (is_null($clientCertSource)) {
+            $this->markTestSkipped('No client cert source found');
+        }
+        $creds = $clientCertSource();
+        $this->assertTrue(is_string($creds));
+        $this->assertContains('-----BEGIN CERTIFICATE-----', $creds);
+        $this->assertContains('-----BEGIN PRIVATE KEY-----', $creds);
+    }
+
+    /**
+     * @runInSeparateProcess
+     */
+    public function testDefaultClientCertSourceInvalidCmdThrowsException()
```

Comment:
Not sure if at some point the original intent was to throw an exception, but I do like that approach. If we keep as is, we should update the method here.

```suggestion
    public function testDefaultClientCertSourceInvalidCmdReturnsNull()
```

Reply:
I agree throwing an exception is better. I can't remember if this was to ensure we don't break customers, or to be consistent with other languages, but I will update it to throw an exception as that's a clearer UX.

Reply:
Updated to throw an exception if the command errors